lib/unlock-helpers.js

```diff
@@ -0,0 +1,175 @@
+      pins[text] = els[e];
+    }
+    for (let i in key) {
+      let pin = key[i];
+      let el = pins[pin];
+      await driver.click(el.ELEMENT);
+    }
+    let el = await driver.findElOrEls("id", "com.android.systemui:id/key_enter", false);
+    await driver.click(el.ELEMENT);
+  } else {
+    for (let i in key) {
+      let pin = key[i];
+      let el = await driver.findElOrEls("id", `com.android.keyguard:id/key${pin}`, false);
+      if (el === null) {
+        throw new Error("Error finding unlock pin button!");
+      }
+      await driver.click(el.ELEMENT);
+    }
+    let el = await driver.findElOrEls("id", "com.android.keyguard:id/key_enter", false);
+    await driver.click(el.ELEMENT);
+  }
+  await sleep(100);
+};
+
+helpers.passwordUnlock = async function (adb, driver, capabilities) {
+  logger.info(`Trying to unlock device using password ${capabilities.unlockKey}`);
+  await helpers.dismissKeyguard(driver, adb);
+  let key = capabilities.unlockKey;
+  // Replace blank spaces with %s
+  key = helpers.encodePassword(key);
+  // Why adb ? It was less flaky
+  await adb.shell(["input", "text", key]);
+  await sleep(100);
+  await adb.shell(["input", "keyevent", KEYCODE_NUMPAD_ENTER]);
+  await sleep(100);
+};
+
+helpers.getPatternKeyPosition = function (key, pos0, piece) {
+  let xPos = (key, x, piece) => {
+    return Math.round(x + ((key % 3) || 3) * piece - piece/2);
+  };
+  let yPos = (key, y, piece) => {
+    return Math.round(y + (Math.ceil(((key % 9) || 9)/3) * piece - piece/2));
+  };
+  return {x: xPos(key, pos0.x, piece), y: yPos(key, pos0.y, piece)};
+};
+
+helpers.patternUnlock = async function (adb, driver, capabilities) {
+  logger.info(`Trying to unlock device using pattern ${capabilities.unlockKey}`);
+  await helpers.dismissKeyguard(driver, adb);
+  let keys = capabilities.unlockKey.split("");
+  /* We set the device pattern buttons as number of a regular phone
+   *  | • • • |     | 1 2 3 |
+   *  | • • • | --> | 4 5 6 |
+   *  | • • • |     | 7 8 9 |
+  */
+  // The pattern view buttons is not seeing by the uiautomator, so we are going to
+  // try clicking on the buttons using the parent view bounds and math.
+  let el;
+  if (await adb.getApiLevel() >= 21) {
+    el = await driver.findElOrEls("id", "com.android.systemui:id/lockPatternView", false);
+  } else {
+    el = await driver.findElOrEls("id", "com.android.keyguard:id/lockPatternView", false);
+  }
+  let pos0 = await driver.getLocation(el.ELEMENT);
+  let size = await driver.getSize(el.ELEMENT);
+  let piece = size.width/3;
+  let actions = [];
+  let lastPos;
+  for (let i in keys) {
+    let keyPos = helpers.getPatternKeyPosition(keys[i], pos0, piece);
+    if (parseInt(i) === 0) {
+      actions.push({action: 'press', options: {element: null, x: keyPos.x, y: keyPos.y}});
+      lastPos = keyPos;
+      continue;
+    }
+    let moveTo = {x:0, y:0};
+    if (keyPos.x > lastPos.x) moveTo.x = piece;
```

Comment:
Please avoid bracketless if for one-liners. See https://github.com/appium/appium/blob/master/docs/en/contributing-to-appium/style-guide-2.0.md for more details